src/ORM/Query.php

```diff
@@ -1404,23 +1402,4 @@ public function isAutoFieldsEnabled(): ?bool
     {
         return $this->_autoFields;
     }
-
-    /**
-     * Decorates the results iterator with MapReduce routines and formatters
-     *
-     * @param \Traversable $result Original results
-     * @return \Cake\Datasource\ResultSetInterface
-     */
-    protected function _decorateResults(Traversable $result): ResultSetInterface
-    {
-        $result = $this->_applyDecorators($result);
-
-        if (!($result instanceof ResultSet) && $this->isBufferedResultsEnabled()) {
-            $class = $this->_decoratorClass();
-            /** @var \Cake\Datasource\ResultSetInterface $result */
-            $result = new $class($result->buffered());
```

Comment:
Is this case never reached by the testsuite?

Reply:
Running the ORM tests after putting a `dd(\Cake\Error\Debugger::trace());` inside this `if` blocks gives:

```php
.................CORE/src/ORM/Query.php (line 1418)
########## DEBUG ##########
'Cake\ORM\Query::_decorateResults() - CORE/src/ORM/Query.php, line 1418
Cake\ORM\Query::_all() - CORE/src/Datasource/QueryTrait.php, line 293
Cake\ORM\Query::all() - CORE/src/ORM/Query.php, line 1072
Cake\ORM\Query::first() - CORE/src/Datasource/QueryTrait.php, line 444
Cake\ORM\Query::firstOrFail() - CORE/src/Datasource/QueryTrait.php, line 455
Cake\Test\TestCase\ORM\Association\BelongsToTest::testAttachToFormatResultsNoDirtyResults() - CORE/tests/TestCase/ORM/Association/BelongsToTest.php, line 470
PHPUnit\Framework\TestCase::runTest() - CORE/vendor/phpunit/phpunit/src/Framework/TestCase.php, line 1415
PHPUnit\Framework\TestCase::runBare() - CORE/vendor/phpunit/phpunit/src/Framework/TestCase.php, line 1035
PHPUnit\Framework\TestResult::run() - CORE/vendor/phpunit/phpunit/src/Framework/TestResult.php, line 691
PHPUnit\Framework\TestCase::run() - CORE/vendor/phpunit/phpunit/src/Framework/TestCase.php, line 763
PHPUnit\Framework\TestSuite::run() - CORE/vendor/phpunit/phpunit/src/Framework/TestSuite.php, line 597
PHPUnit\Framework\TestSuite::run() - CORE/vendor/phpunit/phpunit/src/Framework/TestSuite.php, line 597
PHPUnit\TextUI\TestRunner::doRun() - CORE/vendor/phpunit/phpunit/src/TextUI/TestRunner.php, line 627
PHPUnit\TextUI\Command::run() - CORE/vendor/phpunit/phpunit/src/TextUI/Command.php, line 204
PHPUnit\TextUI\Command::main() - CORE/vendor/phpunit/phpunit/src/TextUI/Command.php, line 163
[main] - CORE/vendor/phpunit/phpunit/phpunit, line 61'
###########################
```

I don't quite understand how this `find()` in this test case triggers this piece of code:

https://github.com/cakephp/cakephp/blob/a990a303853058669234b452a7e9f6a2de11c03a/tests/TestCase/ORM/Association/BelongsToTest.php#L460-L475

Reply:
It's the result formatter set in the `formatted` finder, it will cause `\Cake\Datasource\ResultSetDecorator` to be returned from `QueryTrait::_decorateResults()`.

Reply:
Right, I missed the finder (which has results formatter, causing the resultset to be decorated).

The decorator already wraps a buffered resultset, so piece of code will unwrap the iterators and wrap it again with a `BufferedIterator`.